### PR TITLE
Change "FocusOutput" to be more in-line with "MoveToOutput"

### PR DESCRIPTION
**Pull request by orfeasxyz**
This is a small patch to address https://github.com/labwc/labwc/issues/2099.

### CODE REVIEW

**Review by ahesford — CHANGES_REQUESTED** [review]
Conceptually, I think it would be good to mimic the `MoveToOutput` behavior here, but there is far too much redundancy. The function `output_get_adjacent_output` (which I might call `output_get_adjacent` for brevity) is essentially identical to `view_get_adjacent_output` except for the latter using `view->output` to grab the output. The function `view_get_adjacent_output` should be eliminated in favor of a common function in `src/action.c` that unifies the output selection before calling `desktop_focus_output` or `view_move_to_output`, depending on the action.

Likewise, the `get_wlr_direction` and `opposite_direction` functions from `view.c` should be made public (and probably relocated into `src/common`) for use without duplication. The names of the functions should be made a bit more descriptive when exposing them publicly. (There is similar duplication of `get_wlr_direction` in `src/overlay.c`, but that is slightly different because it aliases `VIEW_EDGE_CENTER` to `WLR_DIRECTION_UP` rather than throwing an error. This is the version you copied here, but you should instead be using the version from `src/view.c`. I'm not sure if we  still need the unique handling in `src/overlay.c`, but we can keep a separate static function if we do.)

Finally, the configuration parsing for both the `MoveToOutput` and `FocusOutput` actions can be deduplicated.

**Comment by orfeasxyz** [comment]
> Conceptually, I think it would be good to mimic the `MoveToOutput` behavior here, but there is far too much redundancy. The function `output_get_adjacent_output` (which I might call `output_get_adjacent` for brevity) is essentially identical to `view_get_adjacent_output` except for the latter using `view->output` to grab the output. The function `view_get_adjacent_output` should be eliminated in favor of a common function in `src/action.c` that unifies the output selection before calling `desktop_focus_output` or `view_move_to_output`, depending on the action.
> 
> Likewise, the `get_wlr_direction` and `opposite_direction` functions from `view.c` should be made public (and probably relocated into `src/common`) for use without duplication. The names of the functions should be made a bit more descriptive when exposing them publicly. (There is similar duplication of `get_wlr_direction` in `src/overlay.c`, but that is slightly different because it aliases `VIEW_EDGE_CENTER` to `WLR_DIRECTION_UP` rather than throwing an error. This is the version you copied here, but you should instead be using the version from `src/view.c`. I'm not sure if we still need the unique handling in `src/overlay.c`, but we can keep a separate static function if we do.)
> 
> Finally, the configuration parsing for both the `MoveToOutput` and `FocusOutput` actions can be deduplicated.

I rewrote the patch following your suggestions.

File: src/output.c [hunk 1]
```diff
@@ -880,6 +881,49 @@ output_nearest_to_cursor(struct server *server)
 		server->seat.cursor->y);
 }
 
+struct output *
+output_get_adjacent(struct output *output, enum view_edge edge, bool wrap)
+{
+	assert(output);
```
Comment (ahesford) [inline src/output.c:887]: If we let output be NULL, we can use `output_get_adjacent` as a full replacement for `view_get_adjacent_output` and just kill that function entirely:
```suggestion
	if (!output) {
		output = output_nearest_to_cursor(server);
	}
	
	if (!output_is_usable(output)) {
		wlr_log(WLR_ERROR,
			"output is not usable, cannot find adjacent output");
		return NULL;
	}
```

File: src/action.c [hunk 2]
```diff
@@ -1060,7 +1052,24 @@ actions_run(struct view *activator, struct server *server,
...
+				if (!target) {
+					/*
+					 * Most likely because we're already on the
+					 * output furthest in the requested direction
+					 * or the output or direction was invalid.
+					 */
+					wlr_log(WLR_DEBUG, "Invalid output");
+					break;
+				}
+				desktop_focus_output(target);
```
Comment (ahesford) [inline src/action.c:1072]: With the change to `output_get_adjacent` suggested below, the entire block for this action (except for the final call to `desktop_focus_output(target)` can be wrapped in a static function in this file, something like
```c
static struct output *
adjacent_output(struct output *output, struct server *server, struct action *action)
{
	const char *output_name = action_get_str(action, "output", NULL);
	struct output *target = NULL;
	if (output_name) {
		target = output_from_name(server, output_name);
	} else {
		enum view_edge edge =
			action_get_int(action, "direction", VIEW_EDGE_INVALID);
		bool wrap = action_get_bool(action, "wrap", false);
		target = output_get_adjacent(output, edge, wrap);
	}

	if (!target) {
		wlr_log(WLR_DEBUG, "Invalid output");
	}

	return target;
}
```

Then, this action can just do
```c
struct output *target = adjacent_output(NULL, server, action);
if (target) {
	desktop_focus_output(target);
}
break;
```
while the `ACTION_TYPE_MOVE_TO_OUTPUT` block can just do
```c
if (!view) {
	break;
}
struct output *target = adjacent_output(NULL, server, action);
if (target) {
	view_move_to_output(view, target);
}
break;
```

**Review by ahesford — CHANGES_REQUESTED** [review]
This is looking a lot better, but I have a few more refinements to recommend.

**Comment by orfeasxyz** [comment]
I did another set of changes as per suggestion.

File: include/labwc.h [hunk 3]
```diff
@@ -537,6 +538,8 @@ struct output *output_from_wlr_output(struct server *server,
 struct output *output_from_name(struct server *server, const char *name);
 struct output *output_nearest_to(struct server *server, int lx, int ly);
 struct output *output_nearest_to_cursor(struct server *server);
+struct output *output_get_adjacent(struct server *server, struct output *output,
+								   enum view_edge edge, bool wrap);
```
Comment (ahesford) [inline include/labwc.h:542]:
```suggestion
/**
 * output_get_adjacent() - get next output, in a given direction, from a given output
 * @output: reference output
 * @edge: direction in which to look for the nearest output
 * @wrap: if true, wrap around at layout edge
 *
 * Note: if output is NULL, the output nearest the cursor will be used as the reference.
 */
struct output *output_get_adjacent(struct server *server,
	struct output *output, enum view_edge edge, bool wrap);
```

File: include/common/direction.h [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#ifndef LABWC_DIRECTION_H
```
Comment (ahesford) [inline include/common/direction.h:1]:
```suggestion
// SPDX-License-Identifier: GPL-2.0-only
#ifndef LABWC_DIRECTION_H
```

File: src/common/direction.c [hunk 5]
```diff
@@ -0,0 +1,42 @@
+#include <assert.h>
```
Comment (ahesford) [inline src/common/direction.c:1]:
```suggestion
// SPDX-License-Identifier: GPL-2.0-only
#include <assert.h>
```

**Review by ahesford — CHANGES_REQUESTED** [review]
With a few tweaks to satisfy the linter, this looks pretty good to me.

**Comment by orfeasxyz** [comment]
If everything looks ok then we can merge it. Also, thank you a lot for your suggestions.

**Comment by orfeasxyz** [comment]
I made the final adjustments but the CodeStyleCheck keeps failing, is this a problem with the indentation?

**Comment by ahesford** [comment]
It seems I led you astray... the license style in the headers should use `/* ... */` comments, while the style in the source files should use `//` comments.

I made this change, along with a couple of other cleanups. (I think your changes broke `view_get_adjacent_output`, and it's obsolete anyway, so I killed it.) Everything should work as expected now. Please take a look at my changes, test that `FocusOutput` and `MoveToOutput` both behave as expected, and I'll squash and merge if you are happy.

**Comment by orfeasxyz** [comment]
With some quick testing everything seems to be working fine. I think we can go ahead and merge.

**Comment by ahesford** [comment]
Thanks!

**Pull request merged**